### ISSUE TITLE
Stabilise `nix hash` and subcommands

### ISSUE THREAD

**Issue by fricklerhandwerk** [NixOS/nix#8876]
This implements https://github.com/NixOS/rfcs/pull/136. This issue is agreed-upon by the https://github.com/orgs/NixOS/teams/nix-team.

Only leave these two commands, collapsing the existing ones into them:

- [x] `nix hash convert`

  ```

  nix hash convert [--hash-algo (sha256|md5|...)] --from base16|...|sri --to (base64|...) <strings>...
  ```
  
  - the current implementation makes `--from` optional, but we are not quite sure how that works. If it doesn't seem robust, we can make it mandatory instead.
  
  - `--algo` is optional but will error if it can't infer from SRI or guess
 - [ ] `nix hash path`:

   ```
   nix hash path --hash-algo (sha256|md5|...) --to (base16|...|sri) --mode nar|flat|git <path>
   ```

   - `algo` is mandatory, any default would be arbitrary
   - collapse modes to `nar` (serialise the directory as a NAR and hash that), `flat` (single file, error on directory), `git` (directory, according to RFC 133)

**Comment by roberth** [comment]
This should also align with a builtin. Such a builtin has been proposed, but does not have a satisfactory design as of yet 
- #7708

**Comment by Ericson2314** [comment]
A comment in https://github.com/NixOS/nix/issues/4354 suggested we rename our base32 to avoid confusion. I missed that before but now support it — docs aren't good enough.

The new CLI is the perfect time to introduce the name change, since the choice of base is really CLI thing (the language and various file and wire formats do not contain the *names* of base to my knowledge).

**Comment by kolloch** [comment]
> A comment in #4354 suggested we rename our base32 to avoid confusion. I missed that before but now support it — docs aren't good enough.
> 
> The new CLI is the perfect time to introduce the name change, since the choice of base is really CLI thing (the language and various file and wire formats do not contain the _names_ of base to my knowledge).

I am not passionate either way except that we should not break existing code and continue accepting `"base32"` where we currently accept it, maybe with the addition of a deprecation notice.

Has this been decided on by https://github.com/orgs/NixOS/teams/nix-team?

**Comment by kolloch** [comment]
Currently, [this is how the internal hash conversion support looks like](https://github.com/NixOS/nix/blob/ac89bb064aeea85a62b82a6daf0ecca7190a28b7/src/libutil/hash.hh#L64):

```cpp
struct Hash { // ...
     static Hash parseAny(std::string_view s, std::optional<HashType> type);
     // ...
}
```

In your proposal, you suggest making `--type` optional but `--from` obligatory (even though you allow the possibility of auto-detecting.

The current code does the following:

1. If the input is an SRI hash, the hash type is contained in the string and accordingly used. 
2. If the input is not an SRI hash, the hash type is required and the hash format is deduced from the length of the input.

I think that is reasonable but it disallows auto-detection of different formats that result in the same length. But even if we introduce new formats resulting in the same length (e.g. alternative base32 encodings), we could still default to the current formats for compatibility.

Therefore, I'd allow skipping `--type` and `--from` for SRI input hashes.

For non-SRI hashes, I'd make `--type` obligatory (as in the internal code currently) and `--from` optional. If `--from` is not given, the default format based on the length is assumed (as now).

**Comment by kolloch** [comment]
I would also make `--to` optional and default to `--to sri`.

Some examples with my suggestions implemented for `nix hash convert`:

```
ash-5.2$ nix hash convert "sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc="
sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc=
bash-5.2$ nix hash convert --to base32 "sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc="
09ymwqf5i9q7d4dm7x4pjjcqqj0qrcp5lnznbh42gfsci5hcbqqm
bash-5.2$ nix hash convert --from base32 --to base32 "sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc="
error: input hash 'sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc=' does not have the expected format '--from base32'
bash-5.2$ nix hash convert --to base32 "sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc="
09ymwqf5i9q7d4dm7x4pjjcqqj0qrcp5lnznbh42gfsci5hcbqqm
bash-5.2$ nix hash convert 09ymwqf5i9q7d4dm7x4pjjcqqj0qrcp5lnznbh42gfsci5hcbqqm
error: hash '09ymwqf5i9q7d4dm7x4pjjcqqj0qrcp5lnznbh42gfsci5hcbqqm' does not include a type, nor is the type otherwise known from context
bash-5.2$ nix hash convert --type sha256 09ymwqf5i9q7d4dm7x4pjjcqqj0qrcp5lnznbh42gfsci5hcbqqm
sha256-FePFYIlMuycIXPZbWi7LGEiMmZSX9FMbaQenWBzm1Sc=
```

Tell me what you think!

**Comment by Ericson2314** [comment]
We now have `hash convert`, but `hash path` doesn't yet do the right thing.

**Comment by cole-h** [comment]
> We now have `nix hash convert`, but `nix hash path` doesn't yet do the right thing.

For the onlookers / those otherwise interested in helping out, could you say more about what `nix hash path` does wrong / doesn't yet do? (And then maybe edit the stabilisation issue to uncheck the box next to `nix hash` and this issue until this has been resolved?)

**Comment by Ericson2314** [comment]
I think `nix hash path` is untouched and nothing like what it says above. `nix hash convert` is close but in another PR `--algo` was disliked and `--hash-algo` was preferred, so we should make that consistent.

**Comment by Ericson2314** [comment]
the `--mode` flags implementations are also too ad-hoc and thus likely to drift out of sync.

**Comment by kolloch** [comment]
You could edit the issue text to indicate that `nix hash convert` is done except for the`—hash-algo` rename.

**Comment by Ericson2314** [comment]
On the flip side, on could argue that `--algo` is fine for a `nix hash` command, whereas `--hash-algo` is needed for other commands.

In any event, #9815 is a draft PR for the rest of this, whatever we decide.

**Comment by cole-h** [comment]
Now that https://github.com/NixOS/nix/pull/9815 has merged, shall we close this, or do we want to wait on https://github.com/NixOS/nix/pull/10021?

**Comment by thufschmitt** [comment]
#10021 is a backwards-compatible addition, let's not be blocked on it
